src/Game/gamePlayReport.cpp

```diff
@@ -0,0 +1,104 @@
+#include "Game/gamePlayReport.h"
+#include "KingSystem/GameData/gdtCommonFlagsUtils.h"
+#include "KingSystem/GameData/gdtManager.h"
+#include "KingSystem/GameData/gdtTriggerParam.h"
+#include "KingSystem/System/ProductReporter.h"
+#include "KingSystem/System/StageInfo.h"
+
+namespace uking {
+
+void PlayReport::reportKorok(const sead::Vector3f& position) {
+    ksys::ProductReporter::getSomeBool();
+    s32 id = ksys::gdt::getFlag_HiddenKorok_Number();
+
+    PlayReport::Report report(sead::SafeString("korok"), 7,
+                              ksys::PlayReportMgr::instance()->getReporter()->getHeap());
+
+    report.addMapType();
+    report.add(sead::SafeString("Id"), u32(id));
+
+    report.addPlayTimes();
+    report.addPosition({position.x, position.z});
+
+    if (ksys::PlayReportMgr::instance()) {
+        auto* reporter = ksys::PlayReportMgr::instance()->getReporter();
+        if (reporter && reporter->isEnabled())
+            reporter->saveReport(&report);
+    }
+}
+
+void PlayReport::reportDungeon(const sead::SafeString& name, const sead::SafeString& event) {
+    ksys::ProductReporter::getSomeBool();
+
+    if (name.findIndex("Remains") == -1 && name.findIndex("Dungeon") == -1 &&
+        name.findIndex("FinalTrial") == -1)
+        return;
+
+    PlayReport::Report report(sead::SafeString("dungeon"), 6,
+                              ksys::PlayReportMgr::instance()->getReporter()->getHeap());
+
+    report.addMapType();
+    report.add(sead::SafeString("Name"), name);
+    report.add(sead::SafeString("Event"), event);
+
+    report.addPlayTimes();
+    if (ksys::PlayReportMgr::instance()) {
+        auto* reporter = ksys::PlayReportMgr::instance()->getReporter();
+        if (reporter && reporter->isEnabled())
+            reporter->saveReport(&report);
+    }
+}
+
+PlayReport::Report::Report(const sead::FixedSafeString<32>& event_id, s32 num_entries,
+                           sead::Heap* heap)
+    : PlayReport(event_id, num_entries, heap) {}
+
+void PlayReport::Report::addPlayTimes() {
+    if (!ksys::gdt::Manager::instance())
+        return;
+
+    if (!ksys::PlayReportMgr::instance())
+        return;
+
+    if (!ksys::PlayReportMgr::instance()->getReporter())
+        return;
+
+    s32 playTime = ksys::PlayReportMgr::instance()->getReporter()->getS32(
+        ksys::PlayReportKey::PlayReport_PlayTime);
+
+    s32 allPlayTime = ksys::PlayReportMgr::instance()->getReporter()->getS32(
+        ksys::PlayReportKey::PlayReport_AllPlayTime);
+
+    add(sead::SafeString("PlayTime"), u32(playTime));
+    add(sead::SafeString("AllPlayTime"), u32(allPlayTime));
+}
+
+void PlayReport::Report::addPosition(const sead::Vector2f& position) {
+    add(sead::SafeString("PosX"), position.x);
+    add(sead::SafeString("PosZ"), position.y);
+}
+
+void PlayReport::Report::addMapType() {
+    if (!ksys::gdt::Manager::instance())
+        return;
+
+    add(sead::SafeString("IsHardMode"), ksys::gdt::getFlag_AoC_HardMode_Enabled());
```

Comment:
consider adding a line break after this line, so that the variable declaration is easier to see